engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrowArrayBuilder.java

```diff
@@ -0,0 +1,242 @@
+package org.enso.interpreter.runtime.data.vector;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.InvalidArrayIndexException;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.interop.UnsupportedTypeException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import com.oracle.truffle.api.nodes.Node;
+import org.enso.interpreter.dsl.Builtin;
+import org.enso.interpreter.node.expression.builtin.interop.syntax.HostValueToEnsoNode;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.data.EnsoObject;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.error.Warning;
+import org.enso.interpreter.runtime.error.WarningsLibrary;
+import org.enso.interpreter.runtime.error.WithWarnings;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+
+@ExportLibrary(InteropLibrary.class)
+@ExportLibrary(TypesLibrary.class)
+@Builtin(
+    pkg = "mutable",
+    stdlibName = "Standard.Base.Data.ArrowArrayBuilder.ArrowArrayBuilder",
+    containsValues = false)
+public abstract class ArrowArrayBuilder implements EnsoObject {
+
+  protected boolean sealed = false;
+
+  public void seal() {
+    sealed = true;
+  }
+
+  @ExportMessage
+  boolean hasArrayElements() {
+    return true;
+  }
+
+  @ExportMessage
+  boolean isArrayElementModifiable(long index) {
+    return true;
+  }
+
+  @ExportMessage
+  boolean isArrayElementReadable(long index) {
+    return true;
+  }
```

Comment:
```suggestion
  @ExportMessage
  boolean isArrayElementReadable(long index) {
    return false;
  }
```

Shouldn't this be `false`?

I think `readArrayElement` should be consistent with `isArrayElementReadable` i.e. `readArrayElement(i)` may fail if and only if `isArrayElementReadable(i) == false`, and the exception to throw should be `UnsupportedOperationException`

Reply:
If we don't implement most of the Array API, should the ArrowArrayBuilder even say it is an array-like (`hasArrayElements = true`)?

IMO it should just be a helper builder that is _not_ an array. It should return the Arrow array and that probably should `hasArrayElements = true`.

But if the builder we barely even implement the array API and do not really rely on it, I'm not sure if we should pretend like it is an array.

IMHO it would be cleaner to have it be a distinct simple API (with `append` that we need).

But that is just my opinion and I may be missing something, so feel free to ignore.

Reply:
delegates to interop